Fix association with extension issues

kamipo opened the pull request:
This fixes the following issues.

* `association_scope` doesn't include `default_scope`. Should use `scope` instead.
* We can't use `method_missing` for customizing existing method.
* We can't use `relation_delegate_class` for sharing extensions. Should extend per association.

kamipo committed:
Fix association with extension issues

This fixes the following issues.

* `association_scope` doesn't include `default_scope`. Should use `scope` instead.
* We can't use `method_missing` for customizing existing method.
* We can't use `relation_delegate_class` for sharing extensions. Should extend per association.

rafaelfranca commented on activerecord/lib/active_record/associations/collection_proxy.rb:
```diff
@@ -28,9 +28,16 @@ module Associations
     # is computed directly through SQL and does not trigger by itself the
     # instantiation of the actual post records.
     class CollectionProxy < Relation
+      @@_extensions = {}
```
Maybe this would create a race condition when defining two extensions to the same association? @matthewd WDYT?

kamipo replied:
Should be here?

```diff
diff --git a/activerecord/lib/active_record/associations/collection_proxy.rb b/activerecord/lib/active_record/associations/collection_proxy.rb
index 750b5b2116..5d6a1e585e 100644
--- a/activerecord/lib/active_record/associations/collection_proxy.rb
+++ b/activerecord/lib/active_record/associations/collection_proxy.rb
@@ -28,14 +28,16 @@ module Associations
     # is computed directly through SQL and does not trigger by itself the
     # instantiation of the actual post records.
     class CollectionProxy < Relation
-      @@_extensions = {}
+      @@_extensions = {}.extend(Mutex_m)
 
       def initialize(klass, association) #:nodoc:
         @association = association
         super klass, klass.arel_table, klass.predicate_builder
 
-        extensions = @@_extensions[association.reflection] ||=
-          association.method(:scope).super_method.().extensions
+        reflection = association.reflection
+        extensions = @@_extensions[reflection] ||= @@_extensions.synchronize do
+          @@_extensions[reflection] ||= association.method(:scope).super_method.().extensions
+        end
 
         extend(*extensions) if extensions.any?
       end
```

avokhmin replied:
@kamipo , looks good for me, @rafaelfranca ?

k41n replied:
+1

rafaelfranca replied:
We could use concurrent Map here?

kamipo committed:
Cache the association proxy object

Some third party modules expects that association returns same proxy
object each time (e.g. for stubbing collection methods:
https://github.com/rspec/rspec-rails/issues/1817).
So I decided that cache the proxy object and reset scope in the proxy
object each time.

Related context: https://github.com/rails/rails/commit/c86a32d7451c5d901620ac58630460915292f88b#commitcomment-2784312

rafaelfranca commented on activerecord/lib/active_record/associations/collection_proxy.rb:
```diff
@@ -28,9 +28,16 @@ module Associations
         super klass, klass.arel_table, klass.predicate_builder
+
+        extensions = @@_extensions[association.reflection] ||=
+          association.method(:scope).super_method.().extensions
```
The fact we need to call the super methods smell like we have something wrong in the inheritance chain. Can we refactor the code to avoid this?

rafaelfranca commented on activerecord/lib/active_record/associations/collection_proxy.rb:
```diff
@@ -1106,10 +1112,14 @@ def reload
+
+      def reset_scope # :nodoc:
+        @scope = nil
+        method(:reset).super_method.()
```
Same here about `super_method`

kamipo commented on activerecord/lib/active_record/associations/collection_proxy.rb:
```diff
@@ -31,6 +31,9 @@ class CollectionProxy < Relation
         super klass, klass.arel_table, klass.predicate_builder
+
+        extensions = association.extensions
+        extend(*extensions) if extensions.any?
```
I extracted better-performing `association.extensions` as @matthewd's suggestion.
It is 2x slower than caching extensions, but seems acceptable. WDYT?

Caching on:

```
CollectionProxies (1000000):   0.740000   0.000000   0.740000 (  0.737209)
Associations (1000000):   0.110000   0.000000   0.110000 (  0.110163)
```

Caching off:

```
CollectionProxies (1000000):   1.620000   0.000000   1.620000 (  1.629061)
Associations (1000000):   0.110000   0.000000   0.110000 (  0.112852)
```

Benchmark script: https://gist.github.com/kevingriffin/29d7e4774d210a41116c063dfd430de0 in #28656.

matthewd replied:
Looks like we can shave a little bit more: https://gist.github.com/matthewd/d942b8dd3e55f17e2e1850d73b5c3e76

I agree we can then skip caching; it seems close enough, and the cut-down `extensions` method reminds us that the answer *could* vary between owners, given a sufficiently creative scope proc.

kamipo committed:
Refactor `default_scoped` to avoid creating extra relation and merging

kamipo commented on activerecord/lib/active_record/scoping/named.rb:
```diff
@@ -30,13 +30,8 @@ def all
-            relation
-          end
+          scope = relation
+          build_default_scope(scope) || scope
```
I refactored `default_scoped` to avoid creating extra relation and merging.
it will make `klass.all` a little faster.

```
CollectionProxies (1000000):   1.420000   0.010000   1.430000 (  1.517390)
Associations (1000000):   0.120000   0.000000   0.120000 (  0.121722)
```

matthewd replied:
Is it still worth avoiding `klass.all` entirely as per https://github.com/rails/rails/pull/29098#discussion_r118831126? When I measured it, it seemed to get us nearly half way towards the cached speed.

matthewd replied:
Also, is this safe? It sounds like it could change which relation the default_scope block sees, or something.

kamipo replied:
`ActiveRecord::Core#relation` creates new `Relation` instance each time.
https://github.com/rails/rails/blob/v5.1.1/activerecord/lib/active_record/core.rb#L307-L315

And also `build_default_scope` doesn't mutate `base_rel` because using `Relation#merge` (not `Relation#merge!`) for now.

So looks safe I think.

kamipo committed:
Extract `default_extensions` to avoid `klass.all`

As @matthewd's suggestion, if `klass` has no default scope, it will more
faster.

kamipo commented on activerecord/lib/active_record/scoping/named.rb:
```diff
@@ -30,12 +30,15 @@ def all
-            relation.spawn.merge!(scope)
+        def default_extensions # :nodoc:
+          if scope = current_scope || build_default_scope
+            scope.extensions
```
I also measured https://gist.github.com/matthewd/d942b8dd3e55f17e2e1850d73b5c3e76.
It will more faster when `klass` has no default scope.

```
CollectionProxies (1000000):   0.950000   0.000000   0.950000 (  0.948074)
Associations (1000000):   0.110000   0.000000   0.110000 (  0.114421)
```

rafaelfranca reviewed (approved):
LGTM.

@matthewd?
